sys/net/dsm/dsm.c

```diff
@@ -0,0 +1,171 @@
+    uint32_t last_used_sec;
+} dsm_session_t;
+
+static int _find_session(sock_dtls_t *sock, sock_dtls_session_t *to_find,
+                         dsm_session_t **session);
+
+static mutex_t _lock;
+static dsm_session_t _sessions[DTLS_PEER_MAX];
+static uint8_t _available_slots;
+
+void dsm_init(void)
+{
+    mutex_init(&_lock);
+    _available_slots = DTLS_PEER_MAX;
+}
+
+dsm_state_t dsm_store(sock_dtls_t *sock, sock_dtls_session_t *session,
+                      dsm_state_t new_state, bool restore)
+{
+    sock_udp_ep_t ep;
+    dsm_session_t *session_slot = NULL;
+    dsm_state_t prev_state = NO_SPACE;
+    mutex_lock(&_lock);
+
+    ssize_t res = _find_session(sock, session, &session_slot);
+    if (res != -1) {
+        prev_state = session_slot->state;
+        if (session_slot->state != SESSION_STATE_ESTABLISHED) {
+            session_slot->state = new_state;
+        }
+
+        /* no existing session found */
+        if (res == 0) {
+            DEBUG("dsm: no existing session found, storing as new session\n")
+            sock_dtls_session_get_udp_ep(session, &ep);
+            sock_dtls_session_set_udp_ep(&session_slot->session, &ep);
+            session_slot->sock = sock;
+            _available_slots--;
+        }
+
+        /* existing session found and session should be restored */
+        if (res == 1 && restore) {
+            DEBUG("dsm: existing session found, restoring\n")
+            memcpy(session, &session_slot->session, sizeof(sock_dtls_session_t));
+        }
+        session_slot->last_used_sec = (uint32_t)(xtimer_now_usec64() / US_PER_SEC);
+    } else {
+        DEBUG("dsm: no space for session to store\n")
+    }
+
+    mutex_unlock(&_lock);
+    return prev_state;
+}
+
+void dsm_remove(sock_dtls_t *sock, sock_dtls_session_t *session)
+{
+    dsm_session_t *session_slot = NULL;
+    mutex_lock(&_lock);
+    if (_find_session(sock, session, &session_slot) == 1) {
+        session_slot->state = SESSION_STATE_NONE;
+        _available_slots++;
+        DEBUG("dsm: removed session\n");
+    } else {
+        DEBUG("dsm: could not find session to remove, it was probably already removed\n");
+    }
+    mutex_unlock(&_lock);
+}
+
+uint8_t dsm_get_num_available_slots(void)
+{
+    return _available_slots;
+}
+
+uint8_t dsm_get_num_maximum_slots(void)
+{
+    return DTLS_PEER_MAX;
+}
+
+ssize_t dsm_get_oldest_used_session(sock_dtls_t *sock, sock_dtls_session_t *session)
+{
+    int res = -1;
+    dsm_session_t *session_slot = NULL;
+
+    if (dsm_get_num_available_slots() == DTLS_PEER_MAX) {
+        return res;
+    }
+
+    mutex_lock(&_lock);
+    for (uint8_t i=0; i < DTLS_PEER_MAX; i++) {
+        if (_sessions[i].state == SESSION_STATE_ESTABLISHED
+                && _sessions[i].sock == sock) {
+            if (session_slot == NULL
+                || session_slot->last_used_sec > _sessions[i].last_used_sec)
+            {
+                session_slot = &_sessions[i];
+            }
+        }
```

Comment:
style nit: it's possible to reduce the level of nesting here

```suggestion
        if (_sessions[i].state != SESSION_STATE_ESTABLISHED) {
            continue;
        }
        if (_sessions[i].sock != sock) {
            continue;
        }

        if (session_slot == NULL ||
            session_slot->last_used_sec > _sessions[i].last_used_sec) {
            session_slot = &_sessions[i];
        }
```

Reply:
Done.